go/arrow/avro/reader_types.go

```diff
@@ -0,0 +1,892 @@
+		f.appendFunc = func(data interface{}) error {
+			appendStringData(bt, data)
+			return nil
+		}
+	case *array.StructBuilder:
+		// has metadata for Avro Union named types
+		sf := parent.newChild(field.Name, b, field.Metadata)
+		sf.typeName, _ = field.Metadata.GetValue("typeName")
+		sf.isStruct = true
+		// create children
+		for i, f := range field.Type.(*arrow.StructType).Fields() {
+			mapFieldBuilders(bt.FieldBuilder(i), f, sf)
+		}
+		sf.appendFunc = func(data interface{}) error {
+			switch data.(type) {
+			case nil:
+				bt.AppendNull()
+			default:
+				appendStructData(bt, data)
+			}
+			return nil
+		}
+	case *array.Time32Builder:
+		f := parent.newChild(field.Name, b, field.Metadata)
+		f.appendFunc = func(data interface{}) error {
+			appendTime32Data(bt, data)
+			return nil
+		}
+	case *array.Time64Builder:
+		f := parent.newChild(field.Name, b, field.Metadata)
+		f.appendFunc = func(data interface{}) error {
+			appendTime64Data(bt, data)
+			return nil
+		}
+	case *array.TimestampBuilder:
+		f := parent.newChild(field.Name, b, field.Metadata)
+		f.appendFunc = func(data interface{}) error {
+			appendTimestampData(bt, data)
+			return nil
+		}
+	}
+}
+
+func appendBinaryData(b *array.BinaryBuilder, data interface{}) {
+	switch dt := data.(type) {
+	case nil:
+		b.AppendNull()
+	case map[string]any:
+		switch ct := dt["bytes"].(type) {
+		case nil:
+			b.AppendNull()
+		default:
+			b.Append(ct.([]byte))
+		}
+	default:
+		b.Append([]byte(fmt.Sprint(data)))
+	}
+}
+
+func appendBinaryDictData(b *array.BinaryDictionaryBuilder, data interface{}) {
+	switch dt := data.(type) {
+	case nil:
+		b.AppendNull()
+	case string:
+		b.AppendString(dt)
+	case map[string]any:
+		switch dt["string"].(type) {
+		case nil:
+			b.AppendNull()
+		case string:
+			b.AppendString(dt["string"].(string))
+		}
+	}
+}
+
+func appendBoolData(b *array.BooleanBuilder, data interface{}) {
+	switch dt := data.(type) {
+	case nil:
+		b.AppendNull()
+	case bool:
+		b.Append(dt)
+	case map[string]any:
+		switch dt["boolean"].(type) {
+		case nil:
+			b.AppendNull()
+		case bool:
+			b.Append(dt["boolean"].(bool))
+		}
+	}
+}
+
+func appendDate32Data(b *array.Date32Builder, data interface{}) {
+	switch dt := data.(type) {
+	case nil:
+		b.AppendNull()
+	case int32:
+		b.Append(arrow.Date32(dt))
+	case map[string]any:
+		switch dt["int"].(type) {
+		case nil:
+			b.AppendNull()
+		case int32:
+			b.Append(arrow.Date32(dt["int"].(int32)))
+		}
+	}
+}
+
+func appendDecimal128Data(b *array.Decimal128Builder, data interface{}) error {
+	switch dt := data.(type) {
+	case nil:
+		b.AppendNull()
+	case []byte:
+		if len(dt) <= 38 {
+			var intData int64
+			buf := bytes.NewBuffer(dt)
+			err := binary.Read(buf, binary.BigEndian, &intData)
+			if err != nil {
+				return err
+			}
+			b.Append(decimal128.FromI64(intData))
+		} else {
+			var bigIntData big.Int
+			buf := bytes.NewBuffer(dt)
+			err := binary.Read(buf, binary.BigEndian, &bigIntData)
```

Comment:
does this actually work? I didn't think you could use binary.Read with a `big.Int`, but I could be wrong.

Reply:
TBH I didn't have any test files that contain that type so I don't think that code path was ever run; you're right, it should be ```err := binary.Read(buf, binary.BigEndian, &bigIntData.Bytes())``` or something like that. I'll do a bit of research.

Reply:
changed to 
```
	case []byte:
		var bigIntData big.Int
		buf := bytes.NewBuffer(dt)
		b.Append(decimal256.FromBigInt(bigIntData.SetBytes(buf.Bytes())))
```